src/metabase/sync/analyze/classify.clj

```diff
@@ -0,0 +1,110 @@
+   Error handling is handled by `run-classifiers` below, so individual classiers do not need to handle errors themselves.
+
+   In the future, we plan to add more classifiers, including ML ones that run offline."
+  (:require [clojure.data :as data]
+            [clojure.tools.logging :as log]
+            [metabase.models.field :refer [Field]]
+            [metabase.sync
+             [interface :as i]
+             [util :as sync-util]]
+            [metabase.sync.analyze.classifiers
+             [category :as category]
+             [name :as name]
+             [no-preview-display :as no-preview-display]
+             [text-fingerprint :as text-fingerprint]]
+            [metabase.util :as u]
+            [schema.core :as s]
+            [toucan.db :as db]))
+
+;;; +------------------------------------------------------------------------------------------------------------------------+
+;;; |                                             CLASSIFYING INDIVIDUAL FIELDS                                              |
+;;; +------------------------------------------------------------------------------------------------------------------------+
+
+(def ^:private values-that-can-be-set
+  "Columns of Field that classifiers are allowed to set."
+  #{:special_type :preview_display})
+
+(s/defn ^:private ^:always-validate save-field-updates!
+  "Save the updates in UPDATED-FIELD."
+  [original-field :- i/FieldInstance, updated-field :- i/FieldInstance]
+  (let [[_ values-to-set] (data/diff original-field updated-field)]
+    (log/debug (format "Based on classification, updating these values of %s: %s" (sync-util/name-for-logging original-field) values-to-set))
+    ;; Check that we're not trying to set anything that we're not allowed to
+    (doseq [k (keys values-to-set)]
+      (when-not (contains? values-that-can-be-set k)
+        (throw (Exception. (format "Classifiers are not allowed to set the value of %s." k)))))
+    ;; cool, now we should be ok to update the Field
+    (db/update! Field (u/get-id original-field)
+      values-to-set)))
+
+
+(def ^:private classifiers
+  "Various classifier functions available. These should all take two args, a `FieldInstance` and a possibly `nil` `Fingerprint`,
+   and return `FieldInstance` with any inferred property changes, or `nil` if none could be inferred.
+   Order is important!"
+  [name/infer-special-type
+   category/infer-is-category
+   no-preview-display/infer-no-preview-display
+   text-fingerprint/infer-special-type])
+
+(s/defn ^:private ^:always-validate run-classifiers :- i/FieldInstance
+  "Run all the available `classifiers` against FIELD and FINGERPRINT, and return the resulting FIELD with changes
+   decided upon by the classifiers."
+  [field :- i/FieldInstance, fingerprint :- (s/maybe i/Fingerprint)]
+  (loop [field field, [classifier & more] classifiers]
+    (if-not classifier
```

Comment:
There are a few classifiers that do things other than classify special type. For example, we have a classifier that checks and decides whether a Field should be marked `preview_display` = `false`.

So the idea here is each of the classifiers looks at the Field and the fingerprint and then returns the Field with any proposed changes (such as a new special type or different `preview_display` status). As you've noticed currently all of our classifiers are either 0 or 1 so there's no sort of confidence number. I think it makes sense to switch things to work that way as a follow-on so I can get this merged sooner rather than later to unblock @senior